lib/puma/reactor.rb

```diff
@@ -1,394 +1,106 @@
         end
+      else
+        select_loop
       end
     end
 
-    def clear_monitor(mon)
-      @selector.deregister mon.value
-      @monitors.delete mon
-    end
-
-    public
-
-    def run
-      run_internal
-    ensure
-      @trigger.close
-      @ready.close
+    # Add a new IO object to monitor.
+    # The object must respond to #timeout and #timeout_at.
+    # Returns false if the reactor is already shut down.
+    def add(io)
+      @input << io
+      @selector.wakeup
+      true
+    rescue ClosedQueueError
+      false
     end
 
-    def run_in_thread
-      @thread = Thread.new do
-        Puma.set_thread_name "reactor"
-        begin
-          run_internal
-        rescue StandardError => e
-          STDERR.puts "Error in reactor loop escaped: #{e.message} (#{e.class})"
-          STDERR.puts e.backtrace
-          retry
-        ensure
-          @trigger.close
-          @ready.close
-        end
+    # Shutdown the reactor, blocking until the background thread is finished.
+    def shutdown
+      @input.close
+      begin
+        @selector.wakeup
+      rescue IOError # Ignore if selector is already closed
       end
+      @thread.join if @thread
     end
 
-    # The `calculate_sleep` sets the value that the `NIO::Selector#select` will
-    # sleep for in the main reactor loop when no sockets are being written to.
-    #
-    # The values kept in `@timeouts` are sorted so that the first timeout
-    # comes first in the array. When there are no timeouts the default timeout is used.
-    #
-    # Otherwise a sleep value is set that is the same as the amount of time it
-    # would take for the first element to time out.
-    #
-    # If that value is in the past, then a sleep value of zero is used.
-    def calculate_sleep
-      if @timeouts.empty?
-        @sleep_for = DefaultSleepFor
-      else
-        diff = @timeouts.first.value.timeout_at.to_f - Time.now.to_f
+    private
 
-        if diff < 0.0
-          @sleep_for = 0
-        else
-          @sleep_for = diff
+    def select_loop
+      begin
+        until @input.closed? && @input.empty?
+          # Wakeup any registered object that receives incoming data.
+          # Block until the earliest timeout or Selector#wakeup is called.
+          timeout = (earliest = @timeouts.first) && earliest.timeout
+          @selector.select(timeout) {|mon| wakeup!(mon.value)}
+
+          # Wakeup all objects that timed out.
+          timed_out = @timeouts.take_while {|t| t.timeout == 0}
+          timed_out.each(&method(:wakeup!))
+
+          unless @input.empty?
+            register(@input.pop) until @input.empty?
+            @timeouts.sort_by!(&:timeout_at)
+          end
         end
+      rescue StandardError => e
+        STDERR.puts "Error in reactor loop escaped: #{e.message} (#{e.class})"
+        STDERR.puts e.backtrace
+        retry
       end
+      # Wakeup all remaining objects on shutdown.
+      @timeouts.each(&@block.method(:call))
```

Comment:
I think we can pass the block directly, no?

```ruby
      @timeouts.each(&@block)
```

Reply:
Yes of course, silly me! Harmless enough since this has been merged, but worth slipping a fix into a future PR.